File: binary_search_find_median_in_two_sorted_arraies.py
"""
There are two sorted arrays A and B of size m and n respectively. Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
"""


class Solution:
    # @return a float
    def findMedianSortedArrays(self, A, B):
        if (len(A) + len(B)) % 2 == 0:
            return (self.get_kth(A, B, (len(A) + len(B)) / 2) + self.get_kth(A, B, (len(A) + len(B)) / 2 + 1)) / 2.0

        else:
            return self.get_kth(A, B, (len(A) + len(B)) / 2 + 1)

    def get_kth(self, A, B, k):
        if len(A) <= 0:
            return B[k-1]
        if len(B) <= 0:
            return A[k - 1]
        if k <= 1:
            return min(A[0], B[0])
        if len(A) / 2 + len(B) / 2 + 1 >= k:   # drop half of either A or B, take an example to verify the followings. e.g., len(A) = 8, len(B) = 10, k = 9
                                               # then: 4 + 5 + 1 = 10, if k == 9, and B[len(B) / 2] >= A[len(A) / 2], which means, B[5] >= A[4], the first 5 elements in B plus first 3 elements in A already has 9 elements, so the kth element cannot be in the second half of B.
            if B[len(B) / 2] >= A[len(A) / 2]:
                return self.get_kth(A, B[:len(B)/2], k)
            else:
                return self.get_kth(A[:len(A)/2], B, k)
        else:
            if B[len(B) / 2] >= A[len(A) / 2]:
                return self.get_kth(A[len(A) / 2 + 1: ], B, k - (len(A) / 2 + 1))
            else:
                return self.get_kth(A, B[len(B) / 2 + 1: ], k - (len(B) / 2 + 1))

File: C++_version/binary_search_find_median_in_two_sorted_arraies.cpp
/**
"""
There are two sorted arrays A and B of size m and n respectively. Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
"""
*/

#include <iostream>

using namespace std;

class Solution {
public:
    double findMedianSortedArrays(int A[], int m, int B[], int n) {
        if ((m + n) % 2 == 0)
        	return (find_kth(A, m, B, n, (m + n) / 2) + find_kth(A, m, B, n, (m + n) / 2 + 1)) / 2.0;
        else
        	return (find_kth(A, m, B, n, (m + n) / 2 + 1));
    }

    double find_kth(int A[], int m, int B[], int n, int k) {
    	if (m <= 0)
    		return B[k - 1];
    	if (n <= 0)
    		return A[k - 1];
    	if (k <= 1)
    		return min(A[0], B[0]);
    	if (m / 2 + n / 2 + 1 >= k) {
    		if (B[n/2] >= A[m/2])
    			return find_kth(A, m, B, n / 2, k);
    		else
    			return find_kth(A, m / 2, B, n, k);
    	} else {
    		if (B[n/2] >= A[m/2])
    			return (A + m/2, m/2, B, n, k - m/2 + 1);
    		else
    			return (A, m, B + n/2, n/2, k - n/2  + 1);
    	}
    }
};

int main()
{
	Solution sol;
	int A[2] = {1,2};
	int B[1] = {3};
	cout << sol.findMedianSortedArrays(A, 2, B, 1) << endl;
}